packages/material-ui/src/Button/Button.js

```diff
@@ -408,7 +408,10 @@ Button.propTypes = {
   /**
    * The variant to use.
    */
-  variant: PropTypes.oneOf(['contained', 'outlined', 'text']),
+  variant: PropTypes /* @typescript-to-proptypes-ignore */.oneOfType([
+    PropTypes.oneOf(['contained', 'outlined', 'text']),
+    PropTypes.string,
```

Comment:
We don't have to solve this right now but we should probably narrow this down before release by checking if any matcher matched. Right now vanilla users don't get runtime validation if the added the wrong `variant`.

Reply:
Thinking about this, we can actually use the `chainPropTypes` and check if the `props.variant` can be found in the `props.classes`. That way I think we can even get rid of the warnings we have currently in place. What do you think about this?

Reply:
Would looking at `props.classes` still work if we support arbitrary props combination (variant + size + orientation) or if the style engine is styled-components? The new warning you added aimed to mitigate with the problem (while losing the component stack trace)

Reply:
I was thinking as these validations are per prop, we can check for only that prop in the classes? But if I understand your point correctly, it can exists in combination with other props, not just itself, right? In that case I will need to update the warning I added, because currently it is checking only for the variant itself.

Reply:
My assumptions were:

- with styled-components, we won't be able to use `chainPropTypes` to have a warning with the full component trace. We would need a new primitive from React. So if we can't get it with sc, maybe we shouldn't try with JSS in the first place.
- if we want to support `<Typography variant="headline1" size="small" color="brand" />` or any dynamic matching props, we would need to construct the class names dynamically, in this case, maybe `.MuiTypogaphy-headline1SizeSmallColorBrand`.
- for the warning, I think that we should include as much information as possible for the developer to find where it warns. For instance, if we can plot `MuiTypography` + `variant="headline1" />` when the headline1 variant doesn't exist, this help a lot
- I haven't given any thoughts to what should happen with the warning and combination of props. I guess we can start with what's simple (your current warning), add tests for it, and grow from there.

Reply:
>if we want to support <Typography variant="headline1" size="small" color="brand" /> or any dynamic matching props, we would need to construct the class names dynamically, in this case, maybe .MuiTypogaphy-headline1SizeSmallColorBrand.  

We can add this even now, with a simple hook like:

```js
const useThemeVariants = (props, name) => {
  const theme = useTheme();
  const variantsClasses = [];
  if(theme && theme.variants && theme.variants[name]) {
    const themeVariants = theme.variants[name];
    themeVariants.forEach(themeVariant => {
      let isMatch = true;
      Object.keys(themeVariant.props).forEach(key => {
        if(props[key] !== themeVariant.props[key]) {
          isMatch = false;
        }
      });
      if(isMatch) {
        variantsClasses.push(classes[propsToClassKey(themeVariant.props)]);
      }
    })
  }
  return variantsClasses;
}
```
Some of the things are recalcuated two times but I will refactor it if we want to go with something like this.

This will allow us to even support new props 👍 I can clean this up and add it in the PR even now.

------------

For the warnings, we can extend it further, but I would like to see what would be the feedback, as I don't want to add a lot of overload for just warning, the one that is in place now it's pretty simple and I am not sure if we will need more complicated one..

Reply:
Testing more this hook, I definitely think we should add it, as it will guarantee that whatever combination of props the clients will use, will be supported inside the component. @oliviertassinari @eps1lon thoughts?

Reply:
I think that it makes sense. As far as I can think about it, for JSS, we will need to both generate the class names from the theme, and to apply them from the props, I don't see any other way. For styled-components, we will be able to bundle the logic at the same place, under the styled.

Reply:
I've added this in the `withStyles` utility, it matches which classes should we added and appends them in the `classes.root` string. Please take a look and let me know if this makes sense.